device taints and tolerations (KEP 5055)

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Device taints enable DRA drivers or admins to mark device as unusable, which prevents allocating them.
Pods may also get evicted at runtime if a device becomes unusable, depending on the severity of the taint
and whether the claim tolerates the taint.

#### Which issue(s) this PR fixes:

Related-to: https://github.com/kubernetes/enhancements/issues/5055 (initial implementation)

#### Special notes for your reviewer:

Based on https://github.com/kubernetes/kubernetes/pull/130120.

#### Does this PR introduce a user-facing change?
```release-note
DRA: Device taints enable DRA drivers or admins to mark device as unusable, which prevents allocating them. Pods may also get evicted at runtime if a device becomes unusable, depending on the severity of the taint and whether the claim tolerates the taint.
```

nojnhuh committed:
DRA: add ResourceSlice tracker

The purpose of the tracker is to emulate a ResourceSlice informer, including
cache and event handlers. In contrast to that informer, the tracker adds taints
from a DeviceTaint such that they appear in the ResourceSlice device
definition. Code using the tracker doesn't need to care where the taints are
coming from.

The main advantage is that it enables fine-grained reactions to taints that
only affect a few devices, the common case. Without this tracker, the pod
eviction controller would have to sync all pods when any slice or any taint
change.

In the scheduler it avoids re-evaluating the selection criteria repeatedly.
The tracker serves as a cross-pod-scheduling cache.

pohly committed:
DRA device taints: copy taintseviction controller

This is a verbatim copy of the current pkg/controller/taintseviction code,
revision fc268ecd09c83c564868dbca1984b9ff2c654482 (v1.33.0 plus one commit),
minus the TimedWorker helper.

The intent is to modify the code such that it enforces eviction of pods which
use tainted devices.

pohly committed:
DRA: add DRADeviceTaints feature

nojnhuh committed:
DRA scheduler: device taints and tolerations

Thanks to the tracker, the plugin sees all taints directly in the device
definition and can compare it against the tolerations of a request while
trying to find a device for the request.

When the feature is turnedd off, taints are ignored during scheduling.

pohly committed:
DRA: add device taints API

This adds the "DeviceTaint" top-level type to v1alpha3 and related fields to
ResourceSlice and ResourceClaim. It's complete enough bring up an API server
and generate files.

pohly committed:
DRA: generated files for device taints API

pohly committed:
ktesting: fix per-test logging in TContext.Run and WithTB

WithTB was originally defined as "uses the existing logger". But what we want
there and in the newer TContext.Run is the usual per-test logging, now for the
sub-test.

pohly committed:
DRA: add device taint eviction controller

The controller is derived from the node taint eviction controller.
In contrast to that controller it tracks the UID of pods to prevent
deleting the wrong pod when it got replaced.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -297,6 +309,59 @@ type DeviceAttribute struct {
+// The device this taint is attached to has the "effect" on
+// any claim which does not tolerate the taint and, through the claim,
+// to pods using the claim.
+type DeviceTaintAtom struct {
```
This `Atom` suffix exists to solve the conflict with the top-level `DeviceTaint` type. I'm not happy about it and would welcome other suggestions, but ultimately we have conflicting requirements.

We want the top-level type to be nice and simple. `DeviceTaint` is analogous to `DeviceClass`. It now contains exactly one taint, so that also fits.

Then we have:
* `core.Toleration` -> `resource.DeviceToleration`
* `core.Taint` -> `resource.DeviceTaint` - except that that name is taken

I could leave out the `Device` prefix, but most of our other structs have it and there could be confusion when mixing both in the same code base, despite the different package prefix.

everpeace replied:
> This Atom suffix exists to solve the conflict with the top-level DeviceTaint type. I'm not happy about it and would welcome other suggestions, but ultimately we have conflicting requirements.

I'd like to propose `DeviceTaintTerm` instead of `Atom`. I think it's a popular word in the kubernetes API.

> We want the top-level type to be nice and simple. DeviceTaint is analogous to DeviceClass. It now contains exactly one taint, so that also fits.

Or, it's just my quick idea though, how about renaming `DeviceTaint` with `DeviceTaintSet`? Itis because 
- `Set` can express more clearly that it can define a single taint (key,value,effect) on _multiple_ devices in _multiple_ `DeviceSlices` resources, and
- I think taint-ing to single or multiple devices would be a popular use-case, but taint-ing the whole device class is very rare, and

So suffixing `Set` doesn't feel wired for me. Then, we can use the word `"DeviceTaint"` for `key,value,effect,timeAdded` tuples.

WDYT?

pohly replied:
DeviceTaintTerm is better than DeviceTaintAtom (I *really* did not spend much thought on that because I wanted to spark a discussion, not propose a solution...). But it's then still inconsistent with other structs.

I don't like `DeviceTaintSet`. `kubectl get devicetaintset foo` is odd when "foo" isn't a set.

everpeace replied:
> But it's then still inconsistent with other structs.

Yeah, but the root reason is that `DeviceTaint` is already taken...

> kubectl get devicetaintset foo is odd when "foo" isn't a set.

ah, I see... I understand what you mean. But, I also still feel the "device taint" should put on some specific device, not on multiple devices in analogy to the node taint.

pohly replied:
I find it useful to give admins a simple way to taint multiple devices with a single command, for example if they do maintenance of a node that only affects pods using a specific driver. Then the DeviceTaint is not just not for one device.

everpeace replied:
> I find it useful to give admins a simple way to taint multiple devices with a single command, for example, if they do maintenance of a node that only affects pods using a specific driver.

Yes, I fully agree to provide simpler/handy way to taint multiple devices for users. 

> Then the DeviceTaint is not just not for one device.

But, I think we don't need to call the resource as `DeviceTaint`. Although I agree that `DeviceTaint` is a very simple and nice name because user can `kubectl get devicetaint foo` as you wrote, this unfortunately causes

- semantic gap from node taint operations
  - it might increase cognitive load in users side
  - if we introduce `DeviceTaint` as a primitive, users have to recognize
    - "node taint" is always per node
    - "device taint" can be declared per device(in DeviceSlice), OR can be declared with `DeviceTaint` for multiple devices
- `Atom` or `Term` suffix in struct

Thus, I prefer to explore nice naming for the resource that taints multiple devices at once.

Oh, contrary, could we introduce `NodeTaint` resource after this release that can taint multiple nodes? It can resolve the semantic gap between "node taint" and "device taint".

pohly replied:
How about `SharedDeviceTaint`? The description then becomes  *"A shared device taint contains one device taint that applies to one or more devices - in other words, it is shared by different devices. The filter defines which devices are tainted with the device taint."*

johnbelamaric replied:
Other ideas:
* `DeviceTaintPolicy`
* `DeviceTaintCriteria` 
* `DeviceTaintApplier`
* `DeviceTainter`

everpeace replied:
Thank you guys for the nice suggestions!!

> How about SharedDeviceTaint for the top-level type? The description then becomes "A shared device taint contains one device taint that applies to one or more devices - in other words, it is shared by different devices. The filter defines which devices are tainted with the device taint."

`SharedDeviceTaint` sounds just great because this just covers what we want! On the other hand, I also think a bit more general/abstrace name might be convinient in case we will add some enhancements on the resource in the future. For example,

- `AdminDeviceTaint` (by thockin)
- `DeviceTaintPolicy` (by johnblemaric)

pohly replied:
:-1: for `AdminDeviceTaint`: one use case is for an external monitoring solution to taint devices, in which case it wouldn't really be "the admin" which does the tainting. OTOH, there is a precedent.

I like `DeviceTaintPolicy`.

If we think a bit further about potential additions, then fields like "evict pods at a certain rate" align well with "Policy", too. We have other `*Policy` types already (ValidatingAdmissionPolicy, etc.). We don't have `*Criteria/Applier/<something>er`. I think this is indeed my favorite.

sftim replied:
Would we document this within https://kubernetes.io/docs/concepts/policy/ ? It doesn't feel like the other things there (**long lived** policies).

So I suggest DeviceTaintRule or something else that doesn't imply a long lived / infrequently changing thing. These device taints might last only, say, 45 seconds.

pohly replied:
I would document it under "DRA and/or cluster administration".

I don't see any other top-level type which ends with `Rule`, so I am not sure whether that is necessarily better.

sftim replied:
I like making the distinction. SIG Architecture can comment on a good suffix for the new API kind and whether it's helpful to separate long-lived things ("Policy") from potentially short lived things ("Rule" / "Expression" / etc).

I know we haven't made that distinction before, but to me that is only a weak ground for not making it now.

pohly replied:
FWIW, I'm fine either way (`DeviceTaintPolicy` and `DeviceTaintRule`). So is @johnbelamaric  (Slack).

thockin replied:
>👎 for AdminDeviceTaint: one use case is for an external monitoring solution to taint devices, in which case it wouldn't really be "the admin" which does the tainting. OTOH, there is a precedent.

"Admin" doesn't necessarily mean the person, but "administratively".  If you empower yout monitoring to taint devices, it is a co-admin of your cluster.

> I like DeviceTaintPolicy.

I don't love it - FooPolicy is usually something that governs Foo.  It's not governing taints, it is itself a taint.  

> @thockin: can you pick something?

Have you learned nothing in all the time we have worked together?  Don't let me name things.

Honestly, of all the suggestions, I like `AdminDeviceTaint`, but if I have to choose between `...Policy` and `...Rule`, I guess I would choose Rule

johnbelamaric replied:
> I don't love it - FooPolicy is usually something that governs Foo. It's not governing taints, it is itself a taint.

It's not really though, it's a formula/rule/policy to determine which devices are tainted. 

I say we go with Rule, if we need a tie-breaker here.

everpeace replied:
> Would we document this within https://kubernetes.io/docs/concepts/policy/ ? It doesn't feel like the other things there (long lived policies).
> ---
> FooPolicy is usually something that governs Foo. It's not governing taints, it is itself a taint.

Thanks for pointing out. This makes sense.

> but "administratively". If you empower yout monitoring to taint devices, it is a co-admin of your cluster.

I can understand this. However, I also feel such cluster resources are created for administrative purpose in most cases.

So, I prefer DeviceTaintRule, too.

pohly replied:
I think we have `DeviceTaintRule` as winner that either is preferred or at least tolerated by everyone. Thanks everyone!

nojnhuh commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -1106,3 +1258,115 @@ type NetworkDeviceData struct {
+	// The taint that gets applied to matching devices.
+	//
+	// +required
+	Taint DeviceTaintAtom `json:"taints,omitempty" protobuf:"bytes,2,rep,name=taint"`
```
To clarify, we no longer need something like `priority` because taints are intrinsically unique? i.e. multiple taints with the same key/value/effect are equivalent to one taint with that same key/value/effect, and any difference in key/value/effect represents a fundamentally different taint?

pohly replied:
My current thinking is that taints are simply additive. If we have two taints with the same key/value/effect, they both apply. If the effect is NoExecute, the eviction happens relative to the older taint. Therefore no priority is needed. That was more important for patches where two patches really could conflict.

For node taints, there is code which matches taints by key/effect:
https://github.com/kubernetes/kubernetes/blob/a6ad9aa8ada1866e7603c93d0e09b2e0e96826b3/staging/src/k8s.io/api/core/v1/taint.go#L21C1-L25C2

It gets called for example here:
https://github.com/kubernetes/kubernetes/blob/a6ad9aa8ada1866e7603c93d0e09b2e0e96826b3/pkg/util/taints/taints.go#L154-L166

Validation for NodeSpec doesn't seem to enforce uniqueness. That by itself doesn't mean anything (could be an oversight), but I really don't see what enforcing that within a ResourceSlice would provide. We can't enforce it when taints get added through different DeviceTaint objects, so the code should simply deal with it as suggested above.

Same for tolerations. A taint is tolerated if some toleration matches it. It doesn't matter when there are multiple tolerations that are identical - unusual perhaps, but not wrong.

nojnhuh commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -1106,3 +1258,115 @@ type NetworkDeviceData struct {
+}
+
+// DeviceCapacity describes a quantity associated with a device.
+type DeviceCapacity struct {
```
Could we omit this type for now, or is it needed for taints?

pohly replied:
No, we don't need it for v1alpha3. Will remove it.

nojnhuh reviewed (commented):
Took a pass at v1alpha3/types.go, will look through this more once it's ready for review.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -1218,3 +1396,100 @@ type NetworkDeviceData struct {
+	// The taint that gets applied to matching devices.
+	//
+	// +required
+	Taint DeviceTaintAtom
```
Related to https://github.com/kubernetes/kubernetes/pull/130447/files#r1985522226: should this type have the "taint" indirection (current state) or should it embed the struct?

Currently this is:
```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: DeviceTaint
metadata:
  name: example
spec:
  # The entire hardware installation is broken.
  # Evict all pods and don't schedule new ones.
  filter:
    driver: dra.example.com
  taint:
    effect: NoExecute
    key: dra.example.com/health
    timeAdded: ...
    value: "Down"
```

This could be instead (again using sorting as in `kubectl get -o yaml`):

```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: DeviceTaint
metadata:
  name: example
spec:
  # The entire hardware installation is broken.
  # Evict all pods and don't schedule new ones.
  effect: NoExecute
  filter:
    driver: dra.example.com
  key: dra.example.com/health
  timeAdded: ...
  value: "Down"
```

Often, sorting by name makes objects pretty unreadable. Here it's okay, only the `timeAdded` would be nicer at the end in both cases.

pohly replied:
I myself prefer the indirection, event if a taint containing a taint might seem a bit weird. But I think it adds some structure and can be rationalized as "these are the taint attributes". It also matches the "taints" list in a ResourceSlice.

nojnhuh replied:
I think I like the indirection a little more too. That makes it a bit easier to grok at a glance.

everpeace replied:
I'd like to +1 for indirection.

> event if a taint containing a taint might seem a bit weird

Yeah, it's a bit wired tough. But, IF, we could rename this object as `DeviceTaintSet` as I proposed in https://github.com/kubernetes/kubernetes/pull/130447/files#r1989129706, then, it could be less recursive-ish??

```
kind: DeviceTaintSet
metadata:
  name: example
spec:
  filter:
    driver: dra.example.com
  taint:
    effect: NoExecute
    key: dra.example.com/health 
    value: "Down"
```

thockin replied:
Agree with the extra indent

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -1218,3 +1396,100 @@ type NetworkDeviceData struct {
+	// devices match.
+	//
+	// +optional
+	Filter *DeviceTaintFilter
```
Should this be required and non-pointer? Then a "taint for all devices" becomes:

```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: DeviceTaint
metadata:
  name: example
spec:
  # Stop all DRA-related pods.
  filter: {}
  taint:
    key: stop-now
    effect: NoExecute
    timeAdded: ...
```

This makes the less common case of "no filter" stand out a bit more.

nojnhuh replied:
Usually defining empty objects with `{}` in YAML just feels a little weird to me, but I can see the value in requiring that here to prevent users from accidentally evicting all of their DRA-allocated pods. Especially since that use case where users do want to do that is probably pretty marginal.

thockin replied:
I don't know - it feels like handing users a sharp knife with no handle and blaming them if they cut themselves.  I think the default should be safe.

everpeace replied:
> I think the default should be safe.

Agree with this. 

The point is the meaning of "safe", isn't it? 

So, when user creating `DeviceTaint` resource, which could be said to be safer, taints all the device, or taints no device?? I personally prefer "tainting no device" by default.

pohly replied:
> I personally prefer "tainting no device" by default. 

Agreed. We don't even need a new field:
- We use `+required` and  `Filter DeviceTaintFilter`.
- If empty, **nothing** is tainted (change of the default semantic). Should that be allowed at all by validation? It might be useful to allow "test taints" which intentionally don't apply to anything, so I'm leaning towards yes.
- If all devices are meant to be tainted, use CEL. This should be uncommon, so we don't need to make it easy:

```yaml
spec:
  filter:
    selectors:
    - cel:
      expression: true
```

Nice: if some future filter gets added, back-rev clients don't see the new field and don't apply the taint, which is safer.

pohly replied:
OTOH, "empty = nothing selected" is inconsistent with label filter.

The alternative is:
* We use `+optional` and `Filter *DeviceTaintFilter`.
* If **nil*, nothing is tainted.
* If a pointer to an empty struct, all devices are tainted (same as with label filter).

johnbelamaric replied:
Filter -> Selector? Which is the word we use for labels too

then optional doing nothing makes sense

johnbelamaric replied:
Or DeviceTaintSelectorList since there are selectors under it?

johnbelamaric replied:
sorry, I am mixing field names and types

My main point is that "no filter" means nothing gets filtered - everything gets through, so "no filter" should mean everything gets tainted, where as "no selector" could mean nothing gets selected

johnbelamaric replied:
Reading the yaml, you could name use the field name `devices` for the filter, I think. Or `deviceSelector`

pohly replied:
> So https://github.com/kubernetes/kubernetes/pull/130447#discussion_r1993600754 is my preferred approach

This overlapped with John's comments. Let me comment on those:

> Filter -> Selector? Which is the word we use for labels too
> 
> then optional doing nothing makes sense

So `spec.deviceSelector` = `Selector *DeviceTaintSelector` (`+optional`), right? Perhaps renamed to `devices` or `deviceSelector`. Does "empty selector" then mean "all devices", as for label selector ("An empty label selector matches all objects. A null label selector matches no objects.")? I'm assuming yes.

We actually had "selector contains selectors" earlier and renamed the outer one to "filter" to avoid that. With "device selector" at the outer level it looks okay to me:

```yaml
spec:
  deviceSelector:
    selectors:
    - cel:
      expression: <something>
    - cel:
      expression: <else>
```

Seems okay. The wording then is *"A device selector may contain zero or more CEL selectors which all must evaluate to true for a device to be selected.*".

thockin replied:
We have examples and counter-examples all over the codebase.

Services with an empty selector disable selection.

Deployments fail validation.

PersistentVolumeClaim select all volumes.

PodAffinity matches nothing.

The main point I want to make is that you need to document what it means if the user specifies nothing

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -784,6 +899,56 @@ type OpaqueDeviceConfiguration struct {
+	// Exists is equivalent to wildcard for value, so that a ResourceClaim can
+	// tolerate all taints of a particular category.
+	//
+	// +optional
```
+default="Equal" ?

pohly replied:
Yes, probably. I did not include it because the node taint API doesn't have it, but that seems like it was an oversight.

thockin reviewed (commented):
I have to run out but I wanted to send comments on the schema part of API, since this is the first I have seen it.

Overall it seems consistent and not surprising.  We see a lot of analogs between nodes and devices - scheduling, taints, etc.  Volumes are very similar but tend to be more dynamic, so no real need for taints (yet).

sftim commented:
Why isn't DeviceTaint a CRD? It may be obvious, but perhaps we should briefly note that obvious answer.

pohly commented:
> Why isn't DeviceTaint a CRD? It may be obvious, but perhaps we should briefly note that obvious answer.

Because in-tree components depend on it. We don't have a story how to build, install and maintain in-tree APIs with CRDs.

sftim commented:
The API kind may be harder to teach than I'd like. This is because the name DeviceTaint implies you are always tainting exactly one device, but actually you can taint sets of devices.

A different name (yes, I know) could help avoid having to make and restate that point in docs, etc.

pohly commented:
@sftim: please see https://github.com/kubernetes/kubernetes/pull/130447#discussion_r1985522226 for the top-level type naming discussion.

pohly committed:
DRA E2E: tests for device taints

pohly committed:
DRA helper: support dropped fields and TimeAdded defaults

Both the new DeviceTaint.TimeAdded and dropped fields when
the DRADeviceTaints feature is disabled confused the ResourceSlice
controller because what is stored and sent back can be different
from what the controller wants to store.

It's now more lenient regarding TimeAdded (doesn't need to be exact because of
rounding during serialization, only having a value on the server is okay)
and dropped fields (doesn't try to store them again). It also preserves
a server-side TimeAdded when updating slices.

pohly commented:
/test pull-kubernetes-integration
/test pull-kuberneters-verify
/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all

pohly commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit

sftim commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	}
+	sliceTracker, err := resourceslicetracker.StartTracker(ctx, opts)
+	if err != nil {
+		logger.Info("Failed to initialize ResourceSlice patch tracker, stopping to evict pods", "err", err)
```
Is this log entry equivalent?
```suggestion
		logger.Info("Failed to initialize ResourceSlice patch tracker; device taints leading to Pod eviction are now paused", "err", err)
```

I couldn't tell.

pohly replied:
Yes, applied (and "patch" removed - we no longer have patches).

sftim commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	retries = 5
+)
+
+// Controller listens to Taint changes of DRA devices and Toleration changes of ResourceClaims,
+// then deletes Pods which use ResourceClaims that don't tolerate a NoExecute taint.
+// Pods which have already reached a final state (aka terminated) don't need to be deleted.
+//
+// All of the logic which identifies pods which need to be evicted runs in the
+// handle* event handlers. They don't call any blocking method. All the blocking
+// calls happen in a [tainteviction.TimedWorkerQueue], using the context passed to Run.
```
I was expecting a mention of DeviceTaintRule here. Should the comment mention that API?

sftim commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -231,6 +232,32 @@ func startTaintEvictionController(ctx context.Context, controllerContext Control
+		name:     names.DeviceTaintEvictionController,
+		initFunc: startDeviceTaintEvictionController,
+		requiredFeatureGates: []featuregate.Feature{
+			// TODO update app.TestFeatureGatedControllersShouldNotDefineAliases when removing these features.
```
(nit)
```suggestion
			// TODO update app.TestFeatureGatedControllersShouldNotDefineAliases when removing these feature gates.
```

We hope to retain the actual features and graduate them to stable.

pohly replied:
Changed, but only here.

This occurs in a few other places. I am not touching those - or should I?

pohly replied:
Hmm, I thought I had seen it elsewhere, but not anymore. Anyway, the change will be in the next update.

sftim commented on pkg/kubeapiserver/default_storage_factory_builder.go:
```diff
@@ -89,6 +90,7 @@ func NewStorageFactoryConfigEffectiveVersion(effectiveVersion basecompatibility.
 		certificates.Resource("clustertrustbundles").WithVersion("v1beta1"),
 		storage.Resource("volumeattributesclasses").WithVersion("v1beta1"),
 		storagemigration.Resource("storagemigrations").WithVersion("v1alpha1"),
+		resource.Resource("devicetaintrules").WithVersion("v1alpha3"),
```
Why `v1alpha3` and not `v1alpha1`?

pohly replied:
Because that is the current alpha for the resource.k8s.io API group.

Bringing back a v1alpha1 and then adding the type there in addition to v1alpha3 and v1beta1 doesn't make sense.

sftim commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -231,6 +232,32 @@ func startTaintEvictionController(ctx context.Context, controllerContext Control
+		controllerContext.InformerFactory.Core().V1().Pods(),
+		controllerContext.InformerFactory.Resource().V1beta1().ResourceClaims(),
+		controllerContext.InformerFactory.Resource().V1beta1().ResourceSlices(),
+		controllerContext.InformerFactory.Resource().V1alpha3().DeviceTaintRules(),
```
How come it's not `v1alpha1`?

pohly commented:
All tests are passing. I have some minor TODOs where I want to do some benchmarking, but that shouldn't stop people from reviewing.

Before this PR, unit test coverage was:

- `k8s.io/dynamic-resource-allocation/resourceclaim`: 94.4%
- `k8s.io/dynamic-resource-allocation/resourceslice`: 79.3%
- `k8s.io/dynamic-resource-allocation/structured`: 91.5%
- `k8s.io/kubernetes/pkg/apis/resource/validation`: 97.6%
- `k8s.io/kubernetes/pkg/controller/tainteviction`: 81.6%
- `k8s.io/kubernetes/pkg/registry/resource/deviceclass`: 88.2%
- `k8s.io/kubernetes/pkg/registry/resource/deviceclass/storage`: 85.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaim`: 86.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaim/storage`: 73.3%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaimtemplate`: 73.3%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaimtemplate/storage`: 85.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceslice`: 34.5%
- `k8s.io/kubernetes/pkg/registry/resource/resourceslice/storage`: 85.7%
- `k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources`: 79.3%


Now with device taint eviction controller instead of the node eviction controller it is:

- `k8s.io/dynamic-resource-allocation/internal/queue`: 100.0%
- `k8s.io/dynamic-resource-allocation/resourceclaim`: 90.9%
- `k8s.io/dynamic-resource-allocation/resourceslice`: 77.8%
- `k8s.io/dynamic-resource-allocation/resourceslice/tracker`: 65.8%
- `k8s.io/dynamic-resource-allocation/structured`: 91.5%
- `k8s.io/kubernetes/pkg/apis/resource/validation`: 97.5%
- `k8s.io/kubernetes/pkg/controller/devicetainteviction`: 90.3%
- `k8s.io/kubernetes/pkg/registry/resource/deviceclass`: 88.2%
- `k8s.io/kubernetes/pkg/registry/resource/deviceclass/storage`: 85.7%
- `k8s.io/kubernetes/pkg/registry/resource/devicetaintrule`: 71.4%
- `k8s.io/kubernetes/pkg/registry/resource/devicetaintrule/storage`: 85.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaim`: 86.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaim/storage`: 73.3%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaimtemplate`: 73.3%
- `k8s.io/kubernetes/pkg/registry/resource/resourceclaimtemplate/storage`: 85.7%
- `k8s.io/kubernetes/pkg/registry/resource/resourceslice`: 62.8%
- `k8s.io/kubernetes/pkg/registry/resource/resourceslice/storage`: 85.7%
- `k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources`: 79.3%

@dom4ha, @x13n: can you check the scheduler integration? There's a small impact on the autoscaler.

@thockin: will you continue with the API review?

@nojnhuh: can you review the eviction controller and the (slightly revised) ResourceSlice tracker?

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -675,19 +685,12 @@ func validateDeviceAttribute(attribute resource.DeviceAttribute, fldPath *field.
 	if attribute.StringValue != nil {
-		if len(*attribute.StringValue) > resource.DeviceAttributeMaxValueLength {
-			allErrs = append(allErrs, field.TooLong(fldPath.Child("string"), "" /*unused*/, resource.DeviceAttributeMaxValueLength))
-		}
```
This looks like a bad rebase or saving an out-dated file on top of the more recent one. I did not intend to change this... will revert.

pohly replied:
Now I remember... this came from admin-controlled attributes. Splitting out the code into separate functions made it reusable. Probably a worthwhile enhancement by itself, even if not directly required anymore.

pohly commented on pkg/features/kube_features.go:
```diff
@@ -228,6 +235,13 @@ const (
+	//
+	// Enables support the ResourceClaim.status.devices field and for setting this
+	// status from DRA drivers.
+	DRAResourceClaimDeviceStatus featuregate.Feature = "DRAResourceClaimDeviceStatus"
```
This is where it belongs when sorted by name... drive-by fix.

thockin commented on staging/src/k8s.io/api/resource/v1alpha3/register.go:
```diff
@@ -52,6 +52,8 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&ResourceClaimTemplateList{},
 		&ResourceSlice{},
 		&ResourceSliceList{},
+		&DeviceTaintRule{},
```
No equivalent for v1beta2?

pohly replied:
It's an alpha-quality API. I suppose we don't want to make it available in the beta version?

thockin replied:
oh, of course ;)  Sorry.

thockin commented on pkg/registry/resource/devicetaintrule/strategy.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// deviceTaintPolicyStrategy implements behavior for DeviceTaintRule objects
+type deviceTaintPolicyStrategy struct {
```
s/Policy/Rule all over?

pohly replied:
My "sed" expression missed this - sorry! Will fix.

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -893,3 +915,122 @@ func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fl
 }
+
+// ValidateDeviceTaint tests if a DeviceTaintRule object is valid.
+func ValidateDeviceTaint(deviceTaint *resource.DeviceTaintRule) field.ErrorList {
```
ValidateDeviceTaintRule

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -893,3 +915,122 @@ func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fl
+}
+
+// ValidateDeviceTaint tests if a DeviceTaintRule update is valid.
+func ValidateDeviceTaintUpdate(deviceTaint, oldDeviceTaint *resource.DeviceTaintRule) field.ErrorList {
```
...Rule

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -893,3 +915,122 @@ func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fl
+		allErrs = append(allErrs, validateLabelValue(taint.Value, fldPath.Child("value"))...)
+	}
+	switch taint.Effect {
+	case resource.DeviceTaintEffectNoSchedule, resource.DeviceTaintEffectNoExecute:
```
Why not the usual set pattern?

```
if taint.Effect == "" {
    field.Required(...)
} else if !validValues.Has(taint.Effect) {
    field.NotSupported(...)
} else { 
   ....
}
```

pohly replied:
No particular reason. Changed.

FWIW, existing code probably could be updated to follow this pattern a bit more.
I looked for `validValues` and usually found code like this:

https://github.com/kubernetes/kubernetes/blob/ed19dbc79f5bfe479b39d2685247895e735bbf70/pkg/apis/core/validation/validation.go#L3681-L3730

Only the last one uses a set, and it doesn't do the `field.Required`.

thockin replied:
Yeah, the reality is that it is all over the map.  If anyone doesn't see the value of declarative validation, they need to read more of these PRs. :)

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -893,3 +915,122 @@ func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fl
+	if toleration.Key != "" {
+		allErrs = append(allErrs, metav1validation.ValidateLabelName(toleration.Key, fldPath.Child("key"))...)
+	}
+	switch toleration.Operator {
```
same, why not a set?

pohly replied:
Changed.

thockin reviewed (commented):
Schematically still seems OK.  Found a bunch of half-renamed places (sorry!)

nojnhuh reviewed (commented):
Didn't quite make it all the way through today, but I'll continue tomorrow.

nojnhuh commented on test/e2e/dra/dra.go:
```diff
@@ -2006,6 +2089,12 @@ func (b *builder) create(ctx context.Context, objs ...klog.KMetadata) []klog.KMe
+			createdObj, err = b.f.ClientSet.ResourceV1alpha3().DeviceTaintRules().Create(ctx, obj, metav1.CreateOptions{})
+			ginkgo.DeferCleanup(func(ctx context.Context) {
+				err := b.f.ClientSet.ResourceV1alpha3().DeviceTaintRules().Delete(ctx, createdObj.GetName(), metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "delete resource slice patch")
```
```suggestion
				framework.ExpectNoError(err, "delete device taint rule")
```

nojnhuh commented on pkg/controller/devicetainteviction/device_taint_eviction_test.go:
Definitely learned a lot going through these tests!

nojnhuh reviewed (commented):
Okay, all done with my pass!

pohly committed:
metrics testing: add type aliases to avoid direct prometheus imports

In tests it is sometimes unavoidable to use the Prometheus types directly,
for example when writing a custom gatherer which needs to normalize data
before testing it. device_taint_eviction_test.go does this to strip
out unpredictable data in a histogram.

With type aliases in a package that is explicitly meant for tests we
can avoid adding exceptions for such tests to the global exception list.

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction_test.go:
```diff
@@ -0,0 +1,1717 @@
+		return got, err
+	}
+
+	return metricstestutil.GatherAndCompare(prometheus.GathererFunc(gather), strings.NewReader(expectedMetric), names...)
```
I chatted with @dgrisonnet about this on Slack. Very few unit tests cover histograms. If they do, they do with fake delays to get deterministic results. I'm taking a new approach here also with avoiding global registration of the metrics. That's probably why it hasn't come up more before.

I agree, type aliases in testutil avoid this. I'll do it in a separate commit, in case that we don't like that.

pohly replied:
https://github.com/kubernetes/kubernetes/pull/130447/commits/8fc9865e7fba1972fac3d30c17ae223473d3b172 after the most recent force-push.

/assign @dgrisonnet 

For review of that change.

dgrisonnet replied:
lgtm

pohly commented on hack/verify-prometheus-imports.sh:
```diff
@@ -37,6 +37,7 @@ source "${KUBE_ROOT}/hack/lib/util.sh"
 # See: https://github.com/kubernetes/kubernetes/issues/89267
 allowed_prometheus_importers=(
   ./cluster/images/etcd-version-monitor/etcd-version-monitor.go
+  ./pkg/controller/devicetainteviction/device_taint_eviction_test.go
```
Yes, see "metrics testing: add type aliases to avoid direct prometheus imports" commit.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -219,6 +219,18 @@ type BasicDevice struct {
 	// +optional
 	Capacity map[QualifiedName]DeviceCapacity
+
+	// If specified, the device's taints.
```
Maybe?

API docs often focus on just describing the individual field, not how it is used. On the one hand that keeps the docs concise, on the other hand they are also often not useful.

I'm undecided here.

pohly replied:
Done and pushed.

pohly commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -290,6 +302,64 @@ type DeviceAttribute struct {
+	// The taint key to be applied to a device.
+	// Must be a label name.
+	//
+	// +required
+	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
```
I'm not sure whether this `opt` has any effect at all. Removed, and it didn't change the Go code. Perhaps for for other users of the .proto file.

Also fixed for `Effect`.

pohly replied:
Hmm, it didn't change the generated.proto either:
```
type DeviceTaint struct {
	// The taint key to be applied to a device.
	// Must be a label name.
	//
	// +required
	Key string `json:"key" protobuf:"bytes,1,name=key"`
...

=>

// +protobuf.options.(gogoproto.goproto_stringer)=false
message DeviceTaint {
  // The taint key to be applied to a device.
  // Must be a label name.
  //
  // +required
  optional string key = 1;
```

Probably really doesn't matter.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/devicetoleration.go:
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Borderline (?): this is old code, just copied into a new file and slightly changed. Anyway, bumped to 2025.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/devicetoleration_test.go:
```diff
@@ -0,0 +1,125 @@
+	}
+	for _, tc := range testCases {
+		if tolerated := ToleratesTaint(tc.toleration, tc.taint); tc.expectTolerated != tolerated {
+			t.Errorf("[%s] expect %v, got %v: toleration %+v, taint %s", tc.description, tc.expectTolerated, tolerated, tc.toleration, tc.taint)
```
Copy-and-pasted :sweat_smile: 

You are right of course. Let's modernize... done.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker.go:
```diff
@@ -0,0 +1,796 @@
+)
+
+// Tracker maintains a view of ResourceSlice objects with matching
+// DeviceTaints applied. It is backed by informers to process
```
Done, also in tracker_test.go.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker.go:
```diff
@@ -0,0 +1,796 @@
+
+// ListPatchedResourceSlices returns all ResourceSlices in the cluster with
+// modifications from DeviceTaints applied.
+func (t *Tracker) ListPatchedResourceSlices() ([]*resourceapi.ResourceSlice, error) {
```
I think "patched" is fine here.

pohly replied:
It then doesn't quite match how it's called in the scheduler interface, but I think that's okay because the context is different.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker_test.go:
```diff
@@ -0,0 +1,1819 @@
+				assert.Equal(t, "s2", events[4].oldObj.Name)
+			},
+		},
+		// TODO: Should the unit test be able to test the passthrough to the normal informer?
```
Agreed. The passthrough is a very obvious shortcut at the top of every public method. I've removed the TODO.

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	evictionTime *metav1.Time
+}
+
+func (tc *Controller) deletePodHandler(c clientset.Interface, emitEventFunc func(tainteviction.NamespacedObject), controllerName string) func(ctx context.Context, fireAt time.Time, args *tainteviction.WorkArgs) error {
```
Even better: remove the parameter. It was needed for the node eviction controller because that didn't use contextual logging, but here we do and thus don't need to pass the name around.

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	// the new pod and stops retrying.
+	return c.CoreV1().Pods(podRef.Namespace).Delete(ctx, podRef.Name, metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &podRef.UID,
```
That only matters for the node taint eviction controlller: a toleration might get added and then the pod gets deleted before the controller re-evaluates whether it still needs to delete. Except that I think tolerations are immutable, so probably not a problem in practice. What we would need to check here is the ResourceVersion of all objects which caused the decision to evict. The pod content is again not relevant because the claim references are also immutable. It wouldn't be complete anyway, so there is a race between deleting a taint and canceling the eviction.

*Replacing* the pod is something that the node taint controller doesn't handle. Adding the UID is how I close that gap in the device taint controller. Probably not relevant in practice and thus not done for the node taint controller, but I don't like gaps...

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	}
+
+	syncBothClaims := func() {
+		// ReservedFor may have changed. If it did, sync both old and new lists,
+		// otherwise only once (same list).
```
This is a potential optimization. In practice, most claims are used by only a single pod. I think the current code is sufficient.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -1005,6 +1015,24 @@ func (alloc *allocator) allocateDevice(r deviceIndices, device deviceWithID, mus
 
+func allTaintsTolerated(device *draapi.BasicDevice, request requestAccessor) bool {
+	for _, taint := range device.Taints {
+		if !taintTolerated(taint, request) {
```
There's no good way to report that. We could log it, but that log output would not be available to users. We cannot report all devices which were skipped during scheduling because they were tainted, for the same reason why we cannot report all devices which were skipped for other reasons: there will be too many of them.

pohly replied:
A "kubectl allocate-device-for-claim" command might be more tractable. That only relies on API semantics and the necessary code is all under `k8s.io/dynamic-resource-allocation` (intentionally!).

nojnhuh replied:
Sorry, I could have made that more clear. I meant to modify the logic here from "if the taint isn't tolerated" to add "and the taint effect is one which prevents scheduling." Currently, all of the taint effects prevent scheduling so it doesn't make a difference, but it might save some surprises later if a new kind of effect is added which doesn't prevent scheduling. (I don't know what effect that might be, just wondering if we could drop the assumption.)

pohly replied:
I think we can change that when the time comes.

pohly commented on pkg/scheduler/eventhandlers.go:
Thanks for reviewing the commit message! Not many people do.

I want to keep the commit separate and care about the text in each of them, so fixed.

pohly commented:
I think I have address the review comments. Here's the diff of the manual changes: https://github.com/kubernetes/kubernetes/compare/23c9caaeab0bb7f7afaba62764a9c7ba18cebdb4..a935736421bacb86c0096e485c51faa18bfa1606

nojnhuh commented on staging/src/k8s.io/api/resource/v1beta1/types.go:
```diff
@@ -300,6 +312,66 @@ type DeviceAttribute struct {
+	// ^^^
+	//
+	// This field was defined as "It is only written for NoExecute taints." for node taints.
+	// But in practice, Kubernetes never did anything with it (no validatio, no defaulting,
```
I think this same typo also got copy-pasted to pkg/apis/resource/types.go and staging/src/k8s.io/api/resource/v1alpha3/types.go.

johnbelamaric commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,922 @@
+	hasSynced atomic.Int32
+}
+
+type poolID struct {
```
This pool management stuff seems widely useful...maybe next cycle it can be factored out for use by other controllers?

pohly replied:
Of the top of my head I can't think of another controller which needs it exactly like this. Let's wait till we have a specific need and then factor out common code.

The namedspacedobject.go is something that I already needed in two places and IMHO should be alongside the NamespacedName type. But I didn't want to start a discussion around that now, so let's keep it a local type for now.

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	// The pre-allocated slices are small enough to be allocated on
+	// the stack (https://stackoverflow.com/a/69187698/222305).
+	p := tc.pools[poolID]
+	oldDeviceTaints := make([]taintedDevice, 0, 10*resourceapi.ResourceSliceMaxDevices)
```
I still wanted to benchmark this together with the other TODOs around performance. Now I have done that and found that it actually causes more allocations! I think that's because Go needs to initialize the elements on the stack, which then apparently causes other allocations. I've simplified to what I should have started with (slice allocation in getTaintedDevices).

Premature optimization, bad Patrick! No biscuit for you today.

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,917 @@
+	// where something changes trigger additional checks for claims
+	// using them.
+	//
+	// The pre-allocated slices are small enough to be allocated on
+	// the stack (https://stackoverflow.com/a/69187698/222305).
```
Absolutely worth double-checking! See https://github.com/kubernetes/kubernetes/pull/130447#discussion_r2001742024.

johnbelamaric reviewed (commented):
some small things, working my way through this

thockin commented on staging/src/k8s.io/component-base/metrics/testutil/testutil.go:
```diff
@@ -20,7 +20,9 @@ import (
 
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/testutil"
+	dto "github.com/prometheus/client_model/go"
```
What does "dto" mean?

pohly replied:
To be honest, I don't know. But it is how this is typically how it is called:

```
cluster/images/etcd-version-monitor/etcd-version-monitor.go:    dto "github.com/prometheus/client_model/go"
staging/src/k8s.io/component-base/metrics/counter.go:   dto "github.com/prometheus/client_model/go"
staging/src/k8s.io/component-base/metrics/counter_test.go:      dto "github.com/prometheus/client_model/go"
staging/src/k8s.io/component-base/metrics/histogram_test.go:    dto "github.com/prometheus/client_model/go"
staging/src/k8s.io/component-base/metrics/metric.go:    dto "github.com/prometheus/client_model/go"
...
```

thockin replied:
weird, it's something we copied from elsewhere:  https://github.com/prometheus/client_golang/blob/main/prometheus/metric.go#L23

thockin reviewed (commented):
API parts LGTM

/approve for API

pohly commented:
Sorry for the noise, I accidentally pushed some other commits. Fixed.

Jefftree commented:
/triage accepted
/retest

johnbelamaric commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
Tons of work to get here, but it all made these changes super clear an easy, nice.

johnbelamaric commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go:
```diff
@@ -2038,6 +2104,65 @@ func TestAllocator(t *testing.T) {
+				localNodeSelector(node1),
+				deviceAllocationResult(req0, driverA, pool1, device1, false),
+			)},
+		},
```
Couple questions:
1) What is the expected behavior of taints with "All" allocation mode? Should All mode fail if a device is tainted? Probably so.
but...
2) (non-blocking) What is the expected behavior of taints with admin access? In particular, shouldn't a DaemonSet that is monitoring devices via admin access, and using All mode, schedule just fine and get access to all devices? Or should it just be sure to set the right tolerations?

pohly replied:
There's nothing special in both cases.

"All" mode requires that the all devices must be usable. If one is tainted, nothing gets allocated and the pod cannot get scheduled.

Admin access also doesn't bypass this check. If an admin was access to to a tainted device, they need to tolerate the taint.

pohly replied:
Might be worth some test cases... I'll add them.

pohly replied:
Done, all worked as described.

johnbelamaric reviewed (approved):
/approve

dom4ha reviewed (commented):
I haven't not done yet, but from scheduler POV it looks good.

> @dom4ha, @x13n: can you check the scheduler integration? There's a small impact on the autoscaler.

@x13n isn't a change required on the CA side as well then?

thockin commented:
/approve for API

everpeace commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -1005,6 +1015,24 @@ func (alloc *allocator) allocateDevice(r deviceIndices, device deviceWithID, mus
 	}, nil
 }
 
+func allTaintsTolerated(device *draapi.BasicDevice, request requestAccessor) bool {
```
This method is called from the `allocator.allocateDevice`.

https://github.com/kubernetes/kubernetes/blob/a0b3bbd1e3ea4ddfb4b3555a385a27f750217b33/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L962-L968

It seems to take both `NoSchedule` and `NoExecute` taint/tolerations into consideration when allocating. I'm not actually sure, but, should we focus on `NoSchedule` taint only when allocating?

pohly replied:
NoExecute is the stronger one of the two effects and implies NoSchedule.

Suppose the scheduler ignores NoExecute and schedules a point using a device which is tainted with that effect. As soon as the pod is scheduled, the device eviction controller kicks in and deletes the pod because of that NoExecute.

everpeace replied:
Sorry, node taint/toleration scheduler plugin takes both taint effects into considertion when scheduling 🙇 

https://github.com/kubernetes/kubernetes/blob/a0b3bbd1e3ea4ddfb4b3555a385a27f750217b33/pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go#L93-L99

https://github.com/kubernetes/kubernetes/blob/a0b3bbd1e3ea4ddfb4b3555a385a27f750217b33/pkg/scheduler/framework/plugins/helper/taint.go#L21-L28

pohly commented:
/assign @MaciekPytel @gjtempleton @x13n 

For SIG autoscaler contract change approval. This is the last missing approval.

The change in CA should be small (replace ResourceSliceInformer with k8s.io/dynamic-resource-controller/resourceslice/tracker, which intentionally has a similar API), but it is a change that is needed to support taints also in CA.

pohly commented:
> This is the last missing approval.

Well, only technically. An explicit approval from SIG Scheduling would be nice. @dom4ha [reviewed](https://github.com/kubernetes/kubernetes/pull/130447#pullrequestreview-2696074702).

/assign @sanposhiho

sanposhiho commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -956,6 +960,12 @@ func (alloc *allocator) allocateDevice(r deviceIndices, device deviceWithID, mus
+	// Might be tainted, in which case the taint has to be tolerated.
+	// The check is skipped if the feature is disabled.
+	if alloc.deviceTaintsEnabled && !allTaintsTolerated(device.basic, request) {
+		return false, nil, nil
```
Not specifically about this feature though, are we not providing users with a way of knowing why each device is rejected? i.e., if a device is rejected by this feature, how could users notice that?

pohly replied:
This is indeed problematic and also applies to other device selection criteria and more generally to all scheduling decisions. I discussed some ideas [earlier in this PR](https://github.com/kubernetes/kubernetes/pull/130447#discussion_r2000682730).

everpeace replied:
I think the device device allocation is tied to `ResourceClaim` resource. So, How about exposing allocation failure (I imagined "can't tolerate" case only) information on `ResourceClaim` side by events or status??

pohly replied:
There may be hundreds of devices that can't be used to allocate a ResourceClaim. We cannot list all of them and the reason in each case.

A single "cannot allocate" event would be possible, but also not very informative.

x13n commented:
I see the only change in CA is a function rename, which is fine. Approving from CA point of view.

/approve

sanposhiho reviewed (approved):
/lgtm
/approve

for the sig-scheduling area. Leave one question, but not blocking one

dom4ha reviewed (commented):
/lgtm

pohly commented:
/hold

I think we have all the necessary approval and reviews, but let's give everyone involved so far a chance to weigh in case that something was missed.

dgrisonnet commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -0,0 +1,916 @@
+			taintInformer.Informer().HasSynced,
+			classInformer.Informer().HasSynced,
+		},
+		metrics: metrics.Global,
```
What's the reason why you need to use a global? Couldn't you call `metrics.New()` here and then modify the `Register` to take Metrics as receiver?

pohly replied:
This is the constructor called by kube-controller-manager. kube-controller-manager expects all controllers to use global metrics. I didn't want to break that contract.

pohly replied:
For example, while unusual it wouldn't be wrong to call New multiple times for the same name as long as only one of them gets to run or the running instance is stopped.

pohly replied:
I tried, but gave up because it also broke unit testing where New is called multiple times and the metrics names are expected to use the same name for all test cases.

pohly commented:
> I think we have all the necessary approval and reviews, but let's give everyone involved so far a chance to weigh in, in case that something was missed.

If I don't hear otherwise, then I'll lift the hold in an hour.

johnbelamaric commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -313,3 +313,5 @@ func dropDeallocatedStatusDevices(newClaim, oldClaim *resource.ResourceClaim) {
 	}
 }
+
+// TODO: add tests after partitionable devices is merged (code conflict!)
```
@pohly should we sequence this PR after partitionable merges? There are also changes in partitionable that add a `features` struct for the feature gates in allocator.

pohly replied:
@cici37: you can remove this comment as part of your rebase. I already added those tests, as you will see while resolving conflicts.

pohly replied:
I copy-and-pasted as much of your test changes as possible, you should be able to simple add your new test cases.

cici37 replied:
Thanks. Will review while rebasing^^

johnbelamaric commented:
/hold cancel

cici37 commented:
/retest
